File: matrix/C/matrix.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

#define N 512

int matrix[2][N][N];

int main(int argc, char* argv[])
{
    bool print_value = false;
    if(argc == 2)
        print_value = true;
    int seed = 42;
    int sum = 0;
    for(int i = 0; i < 2; i++)
    {
        for(int j = 0; j < N; j++)
        {
            for(int k = 0; k < N; k++)
            {
                matrix[i][j][k] = seed;
                seed *= 25189;
                seed %= 32749;
            }
        }
    }
    for(int i = 0; i < N; i++)
    {
        for(int j = 0; j < N; j++)
        {
            for(int k = 0; k < N; k++)
            {
                sum += matrix[0][i][k] * matrix[1][k][j];
                sum &= 0x0000FFFF;
            }
        }
    }
    if(!print_value)
        return EXIT_SUCCESS;
    printf("%d\n", sum);
    return EXIT_SUCCESS;
}

File: matrix/Rust/matrix.rs
use std::env;

fn main() {
	const N: usize = 512;
	let mut matrix = [[[0; N]; N]; 3];
    let mut seed = 42;
    let mut sum = 0;
	for i in 0..2 {
		for j in 0..N {
			for k in 0..N {
				matrix[i][j][k] = seed;
                seed *= 25189;
                seed %= 32749;
			}
		}
	}
	for i in 0..N {
		for j in 0..N {
			for k in 0..N {
				matrix[2][i][j] += matrix[0][i][k] * matrix[1][k][j];
                matrix[2][i][j] &= 0xFFFF;
			}
			sum += ((i * j) & 0xFF) * matrix[2][i][j];
            sum &= 0xFFFF;
		}
	}
	let args: Vec<_> = env::args().collect();
	if args.len() > 1 {
		println!("{}", sum);
	}
}
